[v3.2] Upgrade Go 1.18→1.19

LukeShu opened the pull request:
## Description

There's a large-ish amount of supporting changes here, mostly due to the fact that we got a `gofmt` change.  As usual, I suggest a commit-by-commit review.

## Related Issues
None.

## Testing
None.

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [x] I made sure to update `CHANGELOG.md`. - no applicable changes

   Remember, the CHANGELOG needs to mention:
    + Any new features
    + Any changes to our included version of Envoy
    + Any non-backward-compatible changes
    + Any deprecations

 - [ ] This is unlikely to impact how Ambassador performs at scale. - things should improve: "Performance improvements for the switch statement vary but can be on the order of 20% faster."

   Remember, things that might have an impact at scale include:
    + Any significant changes in memory use that might require adjusting the memory limits
    + Any significant changes in CPU use that might require adjusting the CPU limits
    + Anything that might change how many replicas users should use
    + Changes that impact data-plane latency/scalability

 - [x] My change is adequately tested. - CI should cover it

   Remember when considering testing:
    + Your change needs to be specifically covered by tests.
       + Tests need to cover all the states where your change is relevant: for example, if you add a behavior that can be enabled or disabled, you'll need tests that cover the enabled case and tests that cover the disabled case. It's not sufficient just to test with the behavior enabled.
    + You also need to make sure that the _entire area being changed_ has adequate test coverage.
       + If existing tests don't actually cover the entire area being changed, add tests.
       + This applies even for aspects of the area that you're not changing – check the test coverage, and improve it if needed!
    + We should lean on the bulk of code being covered by unit tests, but...
    + ... an end-to-end test should cover the integration points

 - [x] I updated `DEVELOPING.md` with any any special dev tricks I had to use to work on this code efficiently. - no tricks

 - [ ] The changes in this PR have been reviewed for security concerns and adherence to security best practices.

ddymko reviewed (changes requested):
@LukeShu 

There are two files that have references to go 1.18 

- /build-aux/prelude_go.mk
- /docker/test-http/Dockerfile

LanceEa commented on _cxx/envoy.mk:
```diff
@@ -273,6 +273,7 @@ $(OSS_HOME)/pkg/api/envoy: $(OSS_HOME)/pkg/api/%: $(OSS_HOME)/_cxx/envoy/build_g
 	      -- {} +; \
 	  find "$$tmpdir" -name '*.bak' -delete; \
 	  mv "$$tmpdir/$*" $@; \
+	  gofmt -w -s ./pkg/api/envoy; \
```
Isn't `./pkg/api/envoy` generated code? I'm guessing the envoy build container hasn't been updated to 1.19 so its probably not formatting in 1.19 doc format?

In general, I guess I would have thought that we would exclude running gofmt on generated code so just curious on this one?

LanceEa replied:
alright thanks for the explanation. I'm cool with this.

LanceEa reviewed (changes requested):
@LukeShu - Awesome thanks for doing this! I'm happy to stay up-to-date with our deps.

I left some general questions/comments to make sure I'm understanding things properly.

The commit message for ccd55680b27b1a4e70b43ff172ee0b2f274203ea was confusing to me. In general, it appears go 1.19 made some changes to how formatting is done which caused a lot of diffs and in general we just need to get the repository to a Go 1.19 friendly state. Which is annoying but better than trying to fight with gofmt...Am interpreting that correctly?

A couple of places missed for upgrading from go 1.18 to Go 1.19:

- `/docker/test-http/Dockerfile` - needs to be bumped to 1.19
- `build-aux/prelude_go.mk` - some references to ensuring 1.18 is still in this file

LukeShu commented on _cxx/envoy.mk:
```diff
@@ -307,6 +307,7 @@ $(OSS_HOME)/pkg/envoy-control-plane: $(OSS_HOME)/_cxx/go-control-plane FORCE
 	    -e 's,github\.com/envoyproxy/go-control-plane/pkg,github.com/emissary-ingress/emissary/v3/pkg/envoy-control-plane,g' \
 	    -e 's,github\.com/envoyproxy/go-control-plane/envoy,github.com/emissary-ingress/emissary/v3/pkg/api/envoy,g' \
 	    -- {} +; \
+	  sed -i.bak -e 's/^package/\n&/' "$$tmpdir/log/log_test.go"; \
```
Some gofmt behavior changed in 1.19 (see https://go.dev/doc/comment), and envoyproxy/go-control-plane (at least the version we pull in) is formatted per 1.18.

LukeShu commented on build-aux/generate.mk:
```diff
@@ -186,7 +186,6 @@ protoc = @echo PROTOC --$1_out=$2 $<; mkdir -p $2 && $(tools/protoc) \
 
 # proto_path is a list of where to look for .proto files.
 proto_path += $(OSS_HOME)/api # input files must be within the path
-proto_path += $(OSS_HOME)/vendor # for "k8s.io/..."
```
Correct.

LukeShu commented on build-aux/tools.mk:
```diff
@@ -53,24 +53,26 @@ $(tools.bindir)/%: $(tools.srcdir)/%.py
-tools.main-gomod += $(tools/conversion-gen)  # ensure runtime libraries are consistent
-tools.main-gomod += $(tools/go-mkopensource) # ensure it is consistent with py-mkopensource
-tools.main-gomod += $(tools/kubestatus)      # is actually part of Emissary
+tools.main-gomod += $(tools/controller-gen)     # ensure runtime libraries are consistent
```
I don't think I re-ordered? I just re-lined-up the `#`, since `protoc-gen-go-grpc` is wider than any of the existing entries.

LukeShu commented:
v2:

I've expanded on some commit messages, added a comment to `envoy.mk`, and addressed the 2 remaining 1.18 mentions.

<details><summary>Interdiff:</summary>

```patch
diff --git a/Makefile b/Makefile
index 290325712..ae27d1777 100644
--- a/Makefile
+++ b/Makefile
@@ -10,8 +10,15 @@ include build-aux/tools.mk
 BUILD_ARCH ?= linux/amd64
 
 # Bootstrapping the build env
+#
+# _go-version/deps and _go-version/cmd should mostly only be used via
+# go-version.txt (in generate.mk), but we have declared them early
+# here for bootstrapping the build env.  Don't use them directly (not
+# via go-version.txt) except for bootstrapping.
+_go-version/deps = docker/base-python/Dockerfile
+_go-version/cmd = sed -En 's,.*https://dl\.google\.com/go/go([0-9a-z.-]*)\.linux-amd64\.tar\.gz.*,\1,p' < $(_go-version/deps)
 ifneq ($(MAKECMDGOALS),$(OSS_HOME)/build-aux/go-version.txt)
-  $(_prelude.go.ensure)
+  $(call _prelude.go.ensure,$(shell $(_go-version/cmd)))
   ifneq ($(filter $(shell go env GOROOT),$(subst :, ,$(shell go env GOPATH))),)
     $(error Your $$GOPATH (where *your* Go stuff goes) and $$GOROOT (where Go *itself* is installed) are both set to the same directory ($(shell go env GOROOT)); it is remarkable that it has not blown up catastrophically before now)
   endif
diff --git a/_cxx/envoy.mk b/_cxx/envoy.mk
index 6fe03e9cf..9ca73b739 100644
--- a/_cxx/envoy.mk
+++ b/_cxx/envoy.mk
@@ -273,8 +273,11 @@ $(OSS_HOME)/pkg/api/envoy: $(OSS_HOME)/pkg/api/%: $(OSS_HOME)/_cxx/envoy/build_g
 	      -- {} +; \
 	  find "$$tmpdir" -name '*.bak' -delete; \
 	  mv "$$tmpdir/$*" $@; \
-	  gofmt -w -s ./pkg/api/envoy; \
 	}
+# Envoy's build system still uses an old `protoc-gen-go` that emits
+# code that Go 1.19's `gofmt` isn't happy with.  Even generated code
+# should be gofmt-clean, so gofmt it as a post-processing step.
+	gofmt -w -s ./pkg/api/envoy
 
 # The unmodified go-control-plane
 $(OSS_HOME)/_cxx/go-control-plane: FORCE
diff --git a/build-aux/prelude_go.mk b/build-aux/prelude_go.mk
index bbb8f6954..9547b2ca7 100644
--- a/build-aux/prelude_go.mk
+++ b/build-aux/prelude_go.mk
@@ -80,5 +80,4 @@ _prelude.go.VERSION.HAVE = $(if $(_prelude.go.HAVE),$(call _prelude.go.VERSION.g
 #
 # Building Go programs for use by build-aux
 
-_prelude.go.error_unsupported = $(error This Makefile requires Go '1.18' or newer; you $(if $(_prelude.go.HAVE),have '$(_prelude.go.VERSION)',do not seem to have Go))
-_prelude.go.ensure = $(if $(call _prelude.go.VERSION.HAVE,1.18),,$(_prelude.go.error_unsupported))
+_prelude.go.ensure = $(if $(call _prelude.go.VERSION.HAVE,$1),,$(error This Makefile requires Go '$1' or newer; you $(if $(_prelude.go.HAVE),have '$(_prelude.go.VERSION)',do not seem to have Go)))
diff --git a/docker/test-http/Dockerfile b/docker/test-http/Dockerfile
index ff280ac44..8ee01eb99 100644
--- a/docker/test-http/Dockerfile
+++ b/docker/test-http/Dockerfile
@@ -1,7 +1,7 @@
 # The `test-http` image gets built by `pkg/kubeapply` for use by
 # various kubeapply-templated YAML files.
 
-FROM golang:1.18
+FROM golang:1.19
 COPY httptest.go /usr/local/bin/httptest.go
 RUN go build -o /usr/local/bin/httptest /usr/local/bin/httptest.go
 ENTRYPOINT ["/usr/local/bin/httptest"]
```

</details>

<details><summary>Range-diff:</summary>

```patch
 1:  5a4fdaa6e !  1:  020a4ccc6 make go-mod-tidy: Set the Go version in accordance with go-version.txt
    @@ Metadata
      ## Commit message ##
         make go-mod-tidy: Set the Go version in accordance with go-version.txt
     
    +    The motivation for this is to make the upgrade to Go 1.19 (and future
    +    Go versions) easier.
    +
         Signed-off-by: Luke Shumaker <lukeshu@datawire.io>
     
      ## build-aux/generate.mk ##
 2:  3529ab43f !  2:  34e1a6e49 envoy-control-plane: Don't include the license statement in package docs
    @@ Metadata
      ## Commit message ##
         envoy-control-plane: Don't include the license statement in package docs
     
    -    This is motivated by changes in Go 1.19's `gofmt` that would cause it
    -    to change the formatting of that comment.
    +    Go 1.19's `gofmt` is more prescriptive than previous versions about
    +    how code-blocks in doc-comments are indented (see
    +    https://go.dev/doc/comment).  Because of this, if we were to use Go
    +    1.19's `gofmt` to format `log_test.go` as-is, it would change the
    +    indentation of the Apache License notice.  That's not really a problem
    +    in itself, but it points out a problem to us: that the license notice
    +    is being treated as a doc comment, which it shouldn't be.  So use
    +    `sed` to insert a newline to prevent it from being treated as a doc
    +    comment.
     
         Signed-off-by: Luke Shumaker <lukeshu@datawire.io>
     
 3:  6d6b47800 =  3:  a1a295e75 generate.mk: We no longer consume k8s.io protobufs
 4:  0cd08c98f !  4:  9fa9ef933 tools: Upgrade protoc
    @@ Commit message
         protoc's command-line parsing has gotten more ergonomic, I don't think
         generate.mk needs a fancy `protoc` macro anymore.
     
    +    The motivation of this is to facilitate the upgrade of protoc-gen-go
    +    in a future commit, in order to be on a protoc-gen-go that outputs
    +    code in accordance with Go 1.19's gofmt.
    +
         Signed-off-by: Luke Shumaker <lukeshu@datawire.io>
     
      ## build-aux/generate.mk ##
 5:  3be9b82c7 !  5:  693eb47ad tools: Upgrade protoc-gen-go
    @@ Commit message
         I had to embed UnsafeEchoServiceServer in the kat-server, because gRPC is
         silly.
     
    +    The motivation of this is to get to a version that outputs code that
    +    is formatted in accordance with Go 1.19's gofmt.
    +
         Signed-off-by: Luke Shumaker <lukeshu@datawire.io>
     
      ## DEPENDENCIES.md ##
 6:  ccd55680b =  6:  fcc515d9f Fuss with godoc formatting that Go 1.19 gofmt would mess up
 7:  e51d9f759 <  -:  --------- make generate: Run the envoy protobufs through gofmt
 -:  --------- >  7:  cfda7997f make generate: Run the envoy protobufs through gofmt
 -:  --------- >  8:  b14ae6f52 Makefile: Don't hard-code the required Go version in the bootstrap test
 8:  cd51f47f7 !  9:  a0fc0c3b0 Upgrade Go 1.18.5 → 1.19 [ci-skip]
    @@ docker/base-python/Dockerfile: RUN apk --no-cache add \
      
      RUN curl --fail -L https://storage.googleapis.com/kubernetes-release/release/v1.23.3/bin/linux/amd64/kubectl -o /usr/bin/kubectl && \
        chmod a+x /usr/bin/kubectl
    +
    + ## docker/test-http/Dockerfile ##
    +@@
    + # The `test-http` image gets built by `pkg/kubeapply` for use by
    + # various kubeapply-templated YAML files.
    + 
    +-FROM golang:1.18
    ++FROM golang:1.19
    + COPY httptest.go /usr/local/bin/httptest.go
    + RUN go build -o /usr/local/bin/httptest /usr/local/bin/httptest.go
    + ENTRYPOINT ["/usr/local/bin/httptest"]
 9:  cfd27a7a4 = 10:  249502c0a make go-mod-tidy && make generate && go fmt ./... && make generate
```

</details>

LanceEa reviewed (approved):
Once CI is green then lgtm.

ddymko reviewed (approved):
lgtm
